src/psfmachine/machine.py

```diff
@@ -837,289 +635,197 @@ def build_time_model(self, plot=False, downsample=False, split_time_model=False)
-                    flux_binned[seg_mask].shape
+        # get uncontaminated pixel norm-flux
+        flux, flux_err = np.array(
+            [
+                np.array(
+                    [
+                        self.source_mask.multiply(self.flux[idx]).data,
+                        self.source_mask.multiply(self.flux_err[idx]).data,
+                    ]
                 )
-                res = np.ma.masked_array(res, (~k).reshape(flux_binned[seg_mask].shape))
-                bad_targets = sigma_clip(res, sigma=5).mask
-                bad_targets = (
-                    np.ones(flux_binned[seg_mask].shape, bool) & bad_targets.any(axis=0)
-                ).ravel()
-                # k &= ~sigma_clip(flux_binned.ravel() - A3.dot(time_model_w)).mask
-                k &= ~bad_targets
-            #  we save the weights and pixel mask of each segement for later use
-            seg_time_model_w.append(time_model_w)
-            pix_mask_k.append(k)
-
-        self.seg_splits = splits
-        self.time_model_w = np.asarray(seg_time_model_w)
-        self._time_masked = np.asarray(pix_mask_k, dtype=object)
+                for idx in range(self.nt)
+            ]
+        ).transpose((1, 0, 2))
+        flux_norm = flux / np.nanmean(flux, axis=0)
+        flux_err_norm = flux_err / np.nanmean(flux, axis=0)
+
+        # bindown flux arrays
+        flux_binned_raw = P3.bin_func(flux)
+        flux_binned = P3.bin_func(flux_norm)
+        flux_err_binned = P3.bin_func(flux_err_norm)
+
+        # No saturated pixels, 1e5 is a hardcoded value for Kepler.
+        k = (flux_binned_raw < 1e5).all(axis=0)[None, :] * np.ones(
+            flux_binned_raw.shape, bool
+        )
+        # No faint pixels, 100 is a hardcoded value for Kepler.
+        k &= (flux_binned_raw > 100).all(axis=0)[None, :] * np.ones(
+            flux_binned_raw.shape, bool
+        )
+        # No huge variability
+        k &= (np.abs(flux_binned - 1) < 1).all(axis=0)[None, :] * np.ones(
+            flux_binned.shape, bool
+        )
+        # No nans
+        k &= np.isfinite(flux_binned) & np.isfinite(flux_err_binned)
+        k = np.all(k, axis=0)
+        # combine good-behaved pixels with uncontaminated pixels
+        k &= uncontaminated_pixels
+
+        # iterate to remvoe outliers
+        for count in [0, 1, 2]:
+            # need to add pixel_mask=k
+            P3.fit(flux_norm, flux_err=flux_err_norm, pixel_mask=k)
+            res = flux_binned - P3.matrix.dot(P3.weights).reshape(flux_binned.shape)
+            # res = np.ma.masked_array(res, (~k).reshape(flux_binned.shape))
+            res = np.ma.masked_array(res, ~np.tile(k, flux_binned.shape[0]).T)
+            bad_targets = sigma_clip(res, sigma=5).mask
+            bad_targets = bad_targets.any(axis=0)
+            k &= ~bad_targets
+
+        # bookkeeping
+        self.flux_binned = flux_binned
+        self._time_masked_pix = k
+        self.P3 = P3
```

Comment:
Can we call this `P` instead of `P3`? I think it's easier to read it that way as a ⭐ matrix ⭐